text/0000-Component-Manager-Bounds.md

```diff
@@ -0,0 +1,177 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Component Manager Bounds
+
+## Summary
+
+This is a follow up custom components RFC that outlines how we expose a component's DOM structure and associated semantics.
+
+## Motivation
+Historically, classic Ember components have had a built in escape hatch for doing DOM manipulation via `didInsertElement`. This allowed for components to encapsulate some 3rd party JavaScript library or do some fine grain DOM manipulation. We feel that is ability to drop down to a lower level abstraction makes components powerful.
+
+Today, the custom components are not given direct access to the DOM structure; internally known as `Bounds`. This limitation means that custom components cannot easily encapsulate functionality in the same way that classic Ember components can. Expanding the capabilities of custom components to expose the `Bounds` is in order to increase the usefulness of the abstraction.
+
+## Detailed design
+
+In [RFC#2013](https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md#detailed-design) we outlined the component lifecycle and how a custom component is associated with a manager. This design builds off off of that existing work and concepts.
+
+### `elementHook` capability
+
+The primary way of introducing new functionality to the component manager is through capabilities. To allow custom component managers to announce that they support element related hooks, we will introduce an `elementHook` flag to the capabilities. In doing so, bounds will be dispatched to a `didRenderLayout` method on the manager and `willDestroyLayout` will be called during destruction.
+
+```js
+import { capabilities } from '@ember/component';
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.5', {
+    elementHook: true
+  }),
+
+  // ...
+});
+```
+
+### `didRenderLayout` hook
```

Comment:
I know that this is the name internally in the glimmer-vm, but I don't think its particularly relevant as a name.

Possibly: `didCreateBounds`?

Reply:
Actually, can we avoid creating a new hook here and instead add `bounds` as a second argument to the already existing `didCreateComponent` hook?  `didCreateComponent` is part of the `asyncLifecycleCallbacks` capability, and per the prior RFC is where `didInsertElement` should go. 

The name seems much better, and the intent seems **very much** in line with what we are doing here...

Reply:
I don't mind bike shedding about the name, but I don't think we want to bundle it with `didCreateComponent`. There may be situations that you need one but not the other, and we want to keep things as granular as possible. Plus, bundling the hooks forces their timing to be the same, which we probably don't want (and they are in fact different today).

Reply:
Sounds good, I agree.  Still would prefer another name than `didRenderLayout` it _really_ doesn't feel cohesive with the other component manager method names.